### ISSUE TITLE
[Bug] Nextcloud Encryption breaks with OpenSSL 3.x due to legacy RC4 usage

### ISSUE THREAD

**Issue by MartB** [nextcloud/server#32003]
## ℹ️ Should be fixed by https://github.com/nextcloud/server/pull/36173 (NC26)

https://github.com/nextcloud/server/blob/6fa62e9266d4dc57c34d30392bc066b6e1a9eb2d/apps/encryption/lib/Crypto/Crypt.php#L689

Completely breaks the encryption on any system with the default openssl 3.0 config (legacy ciphers are now disabled).

**Error example if this inevitably starts happening for fedora 36 et. al**
```
OCA\Encryption\Exceptions\MultiKeyDecryptException: multikeydecrypt with share key failed:error:0308010C:digital envelope routines::unsupported
```

**Workaround (from within the distribution openssl.conf)**
```
# Configure as (add or uncomment as needed)

[provider_sect]
default = default_sect
legacy = legacy_sect

[default_sect]
activate = 1

[legacy_sect]
activate = 1
```

**Comment by yasij** [comment]
I ran into this after upgrading to Ubuntu 22.04 LTS with Nextcloud 24rc1. Is there a plan to migrate to a nonlegacy cipher for server-side encryption?

**Comment by lxk3il** [comment]
Same Issue here. 
Users are not able to download their data and get a general server error when trying to download files via webbrowser.
Thanks to @MartB I was able to get access to the files again. 
Log contains the following statements: 
- `OCA\Encryption\Exceptions\MultiKeyDecryptException: multikeydecrypt with share key failed:error:0480006C:PEM routines::no start line`
- `OCA\Encryption\Exceptions\MultiKeyDecryptException: multikeydecrypt with share key failed:error:0308010C:digital envelope routines::unsupported`
- `Sabre\DAV\Exception\ServiceUnavailable: Encryption not ready: multikeydecrypt with share key failed:error:0308010C:digital envelope routines::unsupported`

**Comment by brotkastn** [comment]
Hello,
i upgraded my personal Nextcloud from Ubuntu 21.10 to 22.04 yesterday, and was unable to access my files after the Upgrade.

I can confirm that the changes to the openssl.cnf as suggested by @MartB got everything working again, so thank you for that.

Will this affect every Nextcloud-User with enabled encryption, or just those who have enabled `  'encryption.legacy_format_support' => true,` ?

Thank you for your work, have a nice day,
Michael

**Comment by lum4chi** [comment]
Same problem here, after upgrading Ubuntu from 20.04 to 22.04. 
My `openssl.cnf` file has not the legacy lines mentioned above. So I've tried to comment just the remaining lines but no luck for me.
There is something else I can do? Thanks!

**Comment by MartB** [comment]
@lum4chi Just follow the code for a bit you will eventually end up at the RC4 line i quoted.
I added it this way, so the Nextcloud devs know the direct cause.

They indeed need to be added without the ##, as outlined in the comment from the developers "Uncomment the sections that start with ## below to enable the legacy provider. [...]"

```
[provider_sect]
default = default_sect
legacy = legacy_sect

[default_sect]
activate = 1

[legacy_sect]
activate = 1
```

@brotkastn I never had legacy enabled. This breaking change affects every installation. The cipher needs to be changed in order to get rid of the legacy workaround in openssl.

**Comment by antoniotvr** [comment]
Hi.

@MartB the openssl.cnf that must be modified is that one present in /etc/ssl folder?

Because I have tried to modify this file but encrypted files still not could be decrypted.

Thank you

**Comment by MartB** [comment]
@antoniotvr a safe bet is to refer to your specific distribution manual.  
If you do have php-cli support it is possible to use `php --info | grep "Openssl"`

If this does not work for whatever reason and your php is using the same openssl as your system, running `openssl version -a | grep OPENSSLDIR` should yield a path for you.

For Fedora 36 it is `OPENSSLDIR: "/etc/pki/tls"` and the config resides there.

**Comment by 4nanook** [comment]
Wouldn't it be better for NextCloud to upgrade to a current secure encryption rather than requiring that we enable one that is basically as good as plaintext?

**Comment by MartB** [comment]
> Wouldn't it be better for NextCloud to upgrade to a current secure encryption rather than requiring that we enable one that is basically as good as plaintext?

Exactly and that's the reason I labeled the above as a *workaround* only.
The core devs need to figure out an RC4->X key-migration solution sooner or later. 

RC4 was acceptable in the past, but security and packaging issues like this one, make it impossible to justify continued usage.

**Comment by yahesh** [comment]
A possibility could be to wrap `openssl_seal()` and `openssl_open` so that an alternative RC4 implementation could be used in the short term. I implemented something like that for `openssl_open()` for my Nextcloud Server-Side Encryption [rescue tool](https://github.com/syseleven/nextcloud-tools/blob/master/rescue/decrypt-all-files.php). The solution looks like this:

```php
	// define as a constant to speed up decryptions
	define("REPLACE_RC4_ALGO", checkReplaceRC4Algo());

	function checkReplaceRC4Algo() {
		// with OpenSSL v3 we assume that we have to replace the RC4 algo
		$result = (OPENSSL_VERSION_NUMBER >= 0x30000000);

		if ($result) {
			// maybe someone has re-enabled the legacy support in OpenSSL v3
			$result = (false === openssl_encrypt("test", "rc4", "test", OPENSSL_RAW_DATA, null, $tag, null, 0));
		}

		return $result;
	}

	// hands-down implementation of RC4
	function rc4($data, $secret) {
		$result = false;

		// initialize $state
		$state = [];
		for ($i = 0x00; $i <= 0xFF; $i++) {
			$state[$i] = $i;
		}

		// mix $secret into $state
		$indexA = 0x00;
		$indexB = 0x00;
		for ($i = 0x00; $i <= 0xFF; $i++) {
			$indexB = ($indexB + ord($secret[$indexA]) + $state[$i]) % 0x100;

			$tmp            = $state[$i];
			$state[$i]      = $state[$indexB];
			$state[$indexB] = $tmp;

			$indexA = ($indexA + 0x01) % strlen($secret);
		}

		// decrypt $data with $state
		$indexA = 0x00;
		$indexB = 0x00;
		$result = "";
		for ($i = 0x00; $i < strlen($data); $i++) {
			$indexA = ($indexA + 0x01) % 0x100;
			$indexB = ($state[$indexA] + $indexB) % 0x100;

			$tmp            = $state[$indexA];
			$state[$indexA] = $state[$indexB];
			$state[$indexB] = $tmp;

			$result .= chr(ord($data[$i]) ^ $state[($state[$indexA] + $state[$indexB]) % 0x100]);
		}

		return $result;
	}

	function wrapped_openssl_open($data, &$output, $encrypted_key, $private_key, $cipher_algo, $iv = null) {
		$result = false;

		if ((0 === strcasecmp($cipher_algo, "rc4")) && REPLACE_RC4_ALGO) {
			if (openssl_private_decrypt($encrypted_key, $intermediate, $private_key, OPENSSL_PKCS1_PADDING)) {
				$output = rc4($data, $intermediate);
				$result = (false !== $output);
			}
		} else {
			$result = openssl_open($data, $output, $encrypted_key, $private_key, $cipher_algo, $iv);
		}

		return $result;
	}
```

**Comment by Noob3103** [comment]
The workaround works great. It has some issues in the webpage, but at least I could save my data. For ubuntu server users, the file is in /usr/lib/sslopenssl.cnf
Change it as advised by MartB and you will get access again. 
For the noobs like me -> the ending cnf is correct, I thought they meant that was a typo. 

Please post in this thread in case they fixed the status quo. Won't update or touch it while it ain't totally broken.
Thanks to everyone.

**Comment by hughb8on** [comment]
> The workaround works great. It has some issues in the webpage, but at least I could save my data. For ubuntu server users, the file is in /usr/lib/sslopenssl.cnf Change it as advised by MartB and you will get access again. For the noobs like me -> the ending cnf is correct, I thought they meant that was a typo.

On the two Ubuntu 22.04 servers I've looked at the file is /usr/lib/ssl/openssl.cnf

**Comment by pdolinic** [comment]
I experienced this as well tonight after an upgrade to 22.04 and can confirm the mentioned fix works. Hopefully developers can do something about it.

**Comment by KaeTuuN** [comment]
Since OpenSSL 3.x will soon be the preferred SSL Library in most Distributions this issue should be fixed in the near future I think!

Would be great to hear some comment from a Dev on that Topic!

**Comment by d2x** [comment]
Problem still exists in Nextcloud 25 RC3 requiring the overrides from @MartB 

> 
> **Workaround (from within the distribution openssl.conf)**
> 
> ```
> # Configure as (add or uncomment as needed)
> 
> [provider_sect]
> default = default_sect
> legacy = legacy_sect
> 
> [default_sect]
> activate = 1
> 
> [legacy_sect]
> activate = 1
> ```

**Comment by Noob3103** [comment]
**_This is not helpful in fixing or diagnosing the problem._**
After I changed it according to these posts it is working, but as a beginner I do not remember and have not backed up the original openssl.cnf file. 
The reason for trying that is to test if the issue persists now and in the future. Is there a way to find the default configuration for these settings?
My best guess would be:
```
[provider_sect]
default = default_sect
#legacy = legacy_sect commented out

[default_sect]
activate = 1

#[legacy_sect]
#activate = 0
```
I am dependent on the service I am hosting and have broken my MariaDB to an unfixable degree once, so I had to reformat everything.
Thanks to the community for being so helpful and easy to follow fixes.

**Comment by AkechiShiro** [comment]
@Noob3103 I can confirm you this is not an ARM specific issue, running on x86_64 (on a server) I've encountered the same issue. As long as your OpenSSL version is updated to a certain version, the RC4 is deprecated and breaks the server side encryption on the webview.

**Comment by 4nanook** [comment]
Yep 2.x and later this is the case.

-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
  Eskimo North Linux Friendly Internet Access, Shell Accounts, and Hosting.
    Knowledgeable human assistance, not telephone trees or script readers.
  See our web site: http://www.eskimo.com/ (206) 812-0051 or (800) 246-6874.

**Comment by Noob3103** [comment]
If this is not ARM related, I am confused as to how this issue isn't affecting the businesses hosting it. Running it with outdated software and operating system shouldn't be acceptable for them, or do they have the same or better fixes? Not sure what features are lost with the legacy mode, but I have noticed some restrictions and errors.

**Comment by 4nanook** [comment]
As at least one of those businesses, it has.  I've changed my openssl
configuration to enable RC4 for nextcloud, though I'd much rather see nextcloud
adopt a currently secure protocol.

-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
  Eskimo North Linux Friendly Internet Access, Shell Accounts, and Hosting.
    Knowledgeable human assistance, not telephone trees or script readers.
  See our web site: http://www.eskimo.com/ (206) 812-0051 or (800) 246-6874.

**Comment by gautamkrishnar** [comment]
@Noob3103 ` Is there a way to find the default configuration for these settings?` yes you can check: https://github.com/openssl/openssl/blob/master/apps/openssl.cnf#L56-L72

**Comment by yahesh** [comment]
> A possibility could be to wrap `openssl_seal()` and `openssl_open` so that an alternative RC4 implementation could be used in the short term. I implemented something like that for `openssl_open()` for my Nextcloud Server-Side Encryption [rescue tool](https://github.com/syseleven/nextcloud-tools/blob/master/rescue/decrypt-all-files.php). [...]

I thought it might be interesting for the participants of this issue to know that I developed a fix for it in https://github.com/nextcloud/server/pull/35916.

**Comment by Niffecs** [comment]
Has a solution to the problem been found here?

**Comment by Tokukarin** [comment]
I am surprised that this is still not fixed.

**Comment by enoch85** [comment]
> I am surprised that this is still not fixed. 

It will be in [NC 26](https://github.com/nextcloud/server/pull/35916).

**Comment by cdponcedeleon** [comment]
Is there a way to apply the workaround to Docker? I can't find the specified directories anywhere.

**Comment by Obi-Wahn** [comment]
> Is there a way to apply the workaround to Docker? I can't find the specified directories anywhere.

that would be very helpful!

**Comment by solracsf** [comment]
Should be fixed by https://github.com/nextcloud/server/pull/36173 (NC26)
Feedback welcome.

**Comment by nichogenius** [comment]
For anyone else running PHP via PHP-FPM, you need to restart the PHP-FPM service to reload the openssl config.  My changes to my openssl.cnf weren't taking for some reason - took me a while to realize what was going on.
In my case, the command was: sudo systemctl restart php8.1-fpm.service

**Comment by NaggingDaivy** [comment]
> > Is there a way to apply the workaround to Docker? I can't find the specified directories anywhere.
> 
> that would be very helpful!

That would be very helpful indeed, because there is no workaround for docker image at the time :(

**Comment by Nikki1993** [comment]
> > > Is there a way to apply the workaround to Docker? I can't find the specified directories anywhere.
> > 
> > 
> > that would be very helpful!
> 
> That would be very helpful indeed, because there is no workaround for docker image at the time :(

But there is, you can pass whatever to docker image in this case just have a config locally on your machine that you mount 

I am using unraid and nextcloud linuxserver image so `openssl.cnf` is in `/etc/ssl/` inside the container so to override it I placed my own config under `/mnt/cache/appdata/nextcloud/ssl/` and then told docker to mount that directory in the container to path `etc/ssl/`. Seem to have worked for me, not getting anymore encryption errors.

**Comment by SimmerV** [comment]
> 

I have the same setup but don't have /ssl on the unraid side. Any ideas?

**Comment by MartB** [comment]
Just fyi: I am not deploying Nextcloud anymore, so expect no feedback from me.

From looking at the code in [#36173](https://github.com/nextcloud/server/pull/36173), a temporary workaround has been implemented. 
Good to know that this very same PR paves the way for the usage of a more secure cipher in the future. 

Feel free to close, thanks for contributing everyone!

**Comment by EchoDev** [comment]
I updated to v26 and reverted the OpenSSL changes in my config. 

Nextcloud seems to work fine now with OpenSSL 3.x

**Comment by CluelessTechnologist** [comment]
I had the issue for a few weeks running linuxserver.io's NextCloud docker image. This morning I got a notification there was a upgrade available (v26). After upgrading the issue has disappeared. 
Thanks for fixing the issue! 

The con of using docker images are that they are quite static. Since editing the OpenSSL config seemed hard in my case, I just decided to wait for v26 and it paid off! :+1:
